File: DevExpress VCL 23.2.6 Full Source/VCL/ExpressQuantumGrid/Demos/Delphi/CalculatedFieldsDemo/CalculatedFieldsDemoMain.pas
unit CalculatedFieldsDemoMain;

{$I cxVer.inc}

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  Menus, DB, StdCtrls, dxCore, cxStyles, cxCustomData, cxGraphics, cxFilter,
  cxData, cxDataStorage, cxEdit, cxDBData, cxGridLevel, cxClasses, cxControls,
  cxGridCustomView, cxGridCustomTableView, cxGridTableView, cxGridDBTableView,
  cxGrid, cxDBLookupComboBox, cxEditRepositoryItems, ExtCtrls, BaseForm,
  cxLookAndFeels, cxLookAndFeelPainters, ComCtrls, dxmdaset,
  DBClient, MidasLib, cxNavigator, dxDateRanges, cxGridCardView, cxContainer, dxGallery, dxGalleryControl, cxButtons,
  cxGroupBox, cxCurrencyEdit, cxSpinEdit, ImgList, cxImageList, cxLabel,
  cxDataControllerConditionalFormattingRulesManagerDialog, dxBarBuiltInMenu, cxGridCustomPopupMenu, cxGridPopupMenu;

type
  TfrmMain = class(TfmBaseForm)
    cxGroupBox1: TcxGroupBox;
    cxGroupBox2: TcxGroupBox;
    btnShowExpressionEditor: TcxButton;
    galColumns: TdxGalleryControl;
    dxGalleryControl1Group1: TdxGalleryControlGroup;
    dxGalleryControl1Group1Item1: TdxGalleryControlItem;
    dxGalleryControl1Group1Item2: TdxGalleryControlItem;
    Grid: TcxGrid;
    tvOrders: TcxGridDBTableView;
    tvOrdersOrderID: TcxGridDBColumn;
    tvOrdersProductName: TcxGridDBColumn;
    tvOrdersUnitPrice: TcxGridDBColumn;
    tvOrdersQuantity: TcxGridDBColumn;
    tvOrdersDiscount: TcxGridDBColumn;
    tvOrdersDiscountAmount: TcxGridDBColumn;
    tvOrdersTotal: TcxGridDBColumn;
    GridLevel1: TcxGridLevel;
    cdsProducts: TClientDataSet;
    cdsProductsProductID: TIntegerField;
    cdsProductsProductName: TStringField;
    cdsProductsSupplierID: TIntegerField;
    cdsProductsCategoryID: TIntegerField;
    cdsProductsQuantityPerUnit: TStringField;
    cdsProductsUnitPrice: TFloatField;
    cdsProductsUnitsInStock: TIntegerField;
    cdsProductsUnitsOnOrder: TIntegerField;
    cdsProductsReorderLevel: TIntegerField;
    cdsProductsDiscontinued: TBooleanField;
    cdsProductsEAN13: TStringField;
    mdOrder: TdxMemData;
    mdOrderOrderID: TIntegerField;
    mdOrderProductID: TIntegerField;
    mdOrderUnitPrice: TFloatField;
    mdOrderQuantity: TIntegerField;
    mdOrderDiscount: TFloatField;
    mdOrderOrderDate: TDateTimeField;
    mdOrderProductName: TStringField;
    dsOrder: TDataSource;
    cxImageList1: TcxImageList;
    cxLabel1: TcxLabel;
    cxGridPopupMenu1: TcxGridPopupMenu;
    procedure FormCreate(Sender: TObject);
    procedure btnShowExpressionEditorClick(Sender: TObject);
    procedure ValidateExpressionColumn(
      Sender: TcxCustomGridTableItem; ARecord: TcxCustomGridRecord;
      const AValue: Variant; AData: TcxEditValidateInfo);
  end;

var
  frmMain: TfrmMain;

implementation

uses
  dxSpreadSheetTypes, dxSpreadSheetUtils, AboutDemoForm;

{$R *.dfm}

procedure TfrmMain.btnShowExpressionEditorClick(Sender: TObject);
begin
  if galColumns.Gallery.Groups[0].Items[0].Checked then
    tvOrdersTotal.ShowExpressionEditor
  else
    tvOrdersDiscountAmount.ShowExpressionEditor;
end;

procedure TfrmMain.FormCreate(Sender: TObject);
var
  APath: string;
begin
  APath := ExtractFilePath(Application.ExeName) + '..\..\Data\';
  cdsProducts.LoadFromFile(APath + 'Products2.cds');
  mdOrder.LoadFromBinaryFile(APath + 'Order2.dat');
  tvOrders.DataController.Groups.FullExpand;
end;

procedure TfrmMain.ValidateExpressionColumn(Sender: TcxCustomGridTableItem;
  ARecord: TcxCustomGridRecord; const AValue: Variant; AData: TcxEditValidateInfo);
var
  AErrorCode: Integer;
begin
  if ARecord.IsData then
  begin
    AErrorCode := Sender.GridView.DataController.ErrorCodes[ARecord.RecordIndex, Sender.Index];
    if AErrorCode > 0 then
    begin
      AData.ErrorType := eetError;
      AData.ErrorText := dxSpreadSheetErrorCodeToString(TdxSpreadSheetFormulaErrorCode(AErrorCode));
    end;
  end;
end;

end.

File: DevExpress VCL 23.2.6 Full Source/VCL/ExpressQuantumGrid/Demos/CBuilder/CalculatedFieldsDemo/CalculatedFieldsDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "CalculatedFieldsDemoMain.h"
#include "AboutDemoForm.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxGrid"
#pragma link "cxGridBandedTableView"
#pragma link "cxGridCustomPopupMenu"
#pragma link "cxGridCustomTableView"
#pragma link "cxGridCustomView"
#pragma link "cxGridDBBandedTableView"
#pragma link "cxGridLevel"
#pragma link "cxGridPopupMenu"
#pragma link "cxGridTableView"
#pragma link "cxStyles"
#pragma link "cxLookAndFeels"
#pragma link "cxButtons"
#pragma link "cxContainer"
#pragma link "cxCurrencyEdit"
#pragma link "cxDataStorage"
#pragma link "cxDBData"
#pragma link "cxGridCardView"
#pragma link "cxGridDBTableView"
#pragma link "cxGroupBox"
#pragma link "cxLabel"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxNavigator"
#pragma link "cxSpinEdit"
#pragma link "dxDateRanges"
#pragma link "dxGallery"
#pragma link "dxGalleryControl"
#pragma link "cxImageList"
#pragma link "dxmdaset"
#pragma link "dxSpreadSheetTypes"
#pragma link "dxSpreadSheetUtils"
#pragma resource "*.dfm"
TfrmMain *frmMain;
//---------------------------------------------------------------------------
__fastcall TfrmMain::TfrmMain(TComponent* Owner)
  : TfmBaseForm(Owner)
{
}
//---------------------------------------------------------------------------


void __fastcall TfrmMain::btnShowExpressionEditorClick(TObject *Sender)
{
  if (galColumns->Gallery->Groups->Groups[0]->Items->Items[0]->Checked)
	tvOrdersTotal->ShowExpressionEditor();
  else
	tvOrdersDiscountAmount->ShowExpressionEditor();
}
//---------------------------------------------------------------------------

void __fastcall TfrmMain::FormCreate(TObject *Sender)
{
  String APath = ExtractFilePath(Application->ExeName) + "..\\..\\Data\\";
  cdsProducts->LoadFromFile(APath + "Products2.cds");
  mdOrder->LoadFromBinaryFile(APath + "Order2.dat");

  tvOrders->DataController->Groups->FullExpand();
}
//---------------------------------------------------------------------------

void __fastcall TfrmMain::ValidateExpressionColumn(TcxCustomGridTableItem *Sender,
		  TcxCustomGridRecord *ARecord, const Variant &AValue,
          TcxEditValidateInfo *AData)
{
  if (ARecord->IsData) {
	int AErrorCode = Sender->GridView->DataController->ErrorCodes[ARecord->RecordIndex][Sender->Index];
	if (AErrorCode > 0) {
	  AData->ErrorType = eetError;
	  AData->ErrorText = dxSpreadSheetErrorCodeToString((TdxSpreadSheetFormulaErrorCode)AErrorCode);
	}
  }
}
//---------------------------------------------------------------------------

